File: programmers-coding-test/C++/Level1/키패드 누르기.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

string solution(vector<int> numbers, string hand) {

    string answer = "";
    string key[4] = { "123", "456", "789", "*0#" };
    unordered_map<int, pair<int, int>> hmap;

    int left_row = 3, left_col = 0;
    int right_row = 3, right_col = 2;

    for (int i = 0; i < 4; i++)
        for (int j = 0; j < 3; j++)
            if (isdigit(key[i][j])) hmap[key[i][j] - '0'] = { i, j };  // 숫자들의 좌표를 저장 

    for (int number : numbers)
    {
        // 눌러야 할 숫자의 좌표
        int key_row = hmap[number].first;  
        int key_col = hmap[number].second;

        // 각 손가락들과 키의 거리
        int left_dist = abs(left_row - key_row) + abs(left_col - key_col);
        int right_dist = abs(right_row - key_row) + abs(right_col - key_col);

        if (number == 1 || number == 4 || number == 7)
        {
            left_row = key_row;
            left_col = key_col;
            answer.append("L");
        }
        else if (number == 3 || number == 6 || number == 9)
        {
            right_row = key_row;
            right_col = key_col;
            answer.append("R");
        }
        else
        {
            if (left_dist < right_dist || (left_dist == right_dist && hand == "left"))
            {
                left_row = key_row;
                left_col = key_col;
                answer.append("L");
            }
            else
            {
                right_row = key_row;
                right_col = key_col;
                answer.append("R");
            }
        }
    }

    return answer;
}

File: programmers/Python/Level1/키패드 누르기.py
def get_distance(loc1, loc2):
    return abs(loc1[0] - loc2[0]) + abs(loc1[1] - loc2[1])


def solution(numbers, hand):
    answer = ''
    key_loc = [(3, 1), (0, 0), (0, 1), (0, 2), (1, 0), (1, 1), (1, 2), (2, 0), (2, 1), (2, 2)]
    left_loc, right_loc = (3, 0), (3, 2)

    for number in numbers:
        if number in [1, 4, 7]:
            left_loc = key_loc[number]
            answer += 'L'
        if number in [3, 6, 9]:
            right_loc = key_loc[number]
            answer += 'R'
        if number in [2, 5, 8, 0]:
            left_dist = get_distance(left_loc, key_loc[number])
            right_dist = get_distance(right_loc, key_loc[number])
            if left_dist < right_dist or left_dist == right_dist and hand == 'left':
                left_loc = key_loc[number]
                answer += 'L'
            else:
                right_loc = key_loc[number]
                answer += 'R'
    return answer
